### ISSUE TITLE
[BUG] Docker container does not stop in oneshot mode

### ISSUE THREAD

**Issue by tymmej** [mandarons/icloud-docker#265]
**Describe the bug**
I have configured drive and photos sync to one shot by setting `-1` to interval.
After running docker image I expect that it will finish after syncing once.

To confirm I added `&& echo "end"` to `init.sh`:

```
2024-10-10 06:26:45,572 :: INFO :: root :: sync.py :: 75 :: Syncing drive...
2024-10-10 06:26:45,572 :: INFO :: root :: sync.py :: 77 :: Drive synced
2024-10-10 06:26:45,572 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
2024-10-10 06:26:45,572 :: INFO :: root :: sync.py :: 80 :: Syncing photos...
2024-10-10 06:26:45,572 :: INFO :: root :: sync.py :: 82 :: Photos synced
2024-10-10 06:26:45,572 :: INFO :: root :: config_parser.py :: 98 :: Syncing photos every -1 seconds.
2024-10-10 06:26:45,573 :: INFO :: root :: sync.py :: 131 :: Resyncing at Thu Oct 10 06:26:45 2024 ...
2024-10-10 06:26:45,573 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
2024-10-10 06:26:45,573 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
end
[image still runing]
^C^\%
```

**To Reproduce**

1. Set both `sync_intervals` to `-1`
2. Wait for sync
3. See that docker container is still running

**Expected behavior**
Docker container is stopped after syncing

**Comment by robdejonge** [comment]
This also happens to me. 
```
2024-11-30 15:56:43,719 :: INFO :: root :: sync.py :: 77 :: Drive synced
2024-11-30 15:56:43,719 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
2024-11-30 15:56:43,720 :: INFO :: root :: sync.py :: 129 :: Resyncing at Sat Nov 30 15:56:42 2024 ...
2024-11-30 15:56:43,720 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
```

**Comment by nicx** [comment]
same behaviour here. would be great if the container stops.

**Comment by nicx** [comment]
@mandarons a short addition: same problem when using "retry_login_interval: -1". My expectation would be the docker container stops after the first login failed. but it just does not. is it the same problem? or should I open another bug?

**Comment by mandarons** [comment]
> me problem when using "retry_login_interval: -1". My expectation would be the docker container stops after the first login failed. but it just does not. is it the same problem?

When `retry_login_interval` is set to `-1`, it indicates try login only once. This affects only the login flow retries and not whole application. Hence, it correctly won't cause whole application to exit. 

Setting `sync_interval` to `-1` should cause application exit (which appears to be not happening, hence this bug).

**Comment by nicx** [comment]
@mandarons is it possible to exit the container too when the login fails? otherwise I would not recognize in any way that the sync process is no more working.

**Comment by mandarons** [comment]
Yeah, that's possible. I typically get notifications if logins are failing. But there is no point in keeping the container running if login is failing and `retry_login_interval` is set to `-1`.

Let me see if I can handle this case as well once I figure out solution to this bug.

**Comment by nicx** [comment]
> Yeah, that's possible. I typically get notifications if logins are failing. But there is no point in keeping the container running if login is failing and `retry_login_interval` is set to `-1`.
> 
> Let me see if I can handle this case as well once I figure out solution to this bug.

Any progress to this? It happened to me again: the login failed, and I didn't notice for weeks. :(

**Comment by robdejonge** [comment]
Still happens for me, too.

**Comment by mandarons** [comment]
Please try `main` tag build: `docker pull mandarons/icloud-drive:main` and let me know if it fixes the issue.

**Comment by robdejonge** [comment]
**What I did**: 

- Removed the old container
- Removed the old image
- Removed anything in the `session_data` directory 
- Ensured config file contains `sync_interval: -1`
- Started using docker compose
- Provided password and OTP through `exec` command
- `docker stop icloud-drive`
- `docker start icloud-drive `

**The result**:  
Logs end with 
```
2025-09-24 14:09:31,240 :: INFO :: root :: sync.py :: 77 :: Drive synced
2025-09-24 14:09:31,240 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
2025-09-24 14:09:31,240 :: INFO :: root :: sync.py :: 129 :: Resyncing at Wed Sep 24 14:09:30 2025 ...
2025-09-24 14:09:31,240 :: INFO :: root :: config_parser.py :: 83 :: Syncing drive every -1 seconds.
2025-09-24 14:09:31,240 :: INFO :: root :: sync.py :: 137 :: All configured sync intervals are negative, exiting oneshot mode...
```

Container remains up according to `docker ps`, 
but it does not sync again (at least it as not for >10m).

**Comment by mandarons** [comment]
Please try:

1. Use image: `ghcr.io/mandarons/icloud-docker:pr-346`
2. To login manually, use command `docker exec -it icloud /bin/sh -c "su-exec abc icloud --username=<icloud-username> --session-directory=/config/session_data"` (note no use of `--user=abc` and use of `su-exec abc icloud ...`)
3. Optionally, update your docker run/docker compose to persist `/home/abc/.local` directory. This will persist keyring storage - you will not need to re-enter password to store in keyring on every container recreation!

**Comment by robdejonge** [comment]
Yes! 

What I did:
- Edited my compose yaml to use the image as per point #1 above.
- I set `retry_login_interval: 60` when first starting the container
- `docker compose up`
- `docker exec` command as per your point #2 above, providing the password
- Waited for the sync to complete. Given `sync_interval` was set to `-1`, the container exited as expected after the sync completed as I was hoping. 
- Changed the `retry_login_interval` back to -1 so that I don't wake up to 10 failed login notifications because the MFA has expired. 
- `docker start icloud` to see if it would work as I was hoping, and it did!

So this is excellent! Thank you for fixing this problem! 

If I may ask a follow-up to your suggestion to persist `/home/abc/.local` ... I'd want to persist this inside `/opt/icloud-drive` where I currently store the config.yaml, logs, and session_data. 

1. Can I store logs in /logs instead of /config/logs ... will it create this directory if needed?
2. If I take the lazy route and persist /home/abc/.local into /opt/icloud-drive/keyring/ or something, it means those same files would show up in /config/keyring inside the container. Is that going to be a problem? I'll probably restructure /opt/icloud-drive, but just curious.

**Comment by mandarons** [comment]
> Can I store logs in /logs instead of /config/logs ... will it create this directory if needed?

Not yet. I think, I am hard-coding the log file to be in /config directory. You can try persisting `/config/icloud.log` to some other path (e.g. `/logs/icloud.log`) and see what happens. :)

> If I take the lazy route and persist /home/abc/.local into /opt/icloud-drive/keyring/ or something, it means those same files would show up in /config/keyring inside the container. Is that going to be a problem? I'll probably restructure /opt/icloud-drive, but just curious.

Give it a try and see what happens. :) 

This is what I have in my docker-compose.yaml:
```yaml
    volumes:
      - /etc/timezone:/etc/timezone:ro
      - /etc/localtime:/etc/localtime:ro
      - ${PWD}/icloud:/icloud # icloud data
      - ${PWD}/.local:/home/abc/.local # keyring file
      - ${PWD}/config:/config # Must contain config.yaml
 ```

**Comment by robdejonge** [comment]
I've taken the non-lazy route, moved /opt/icloud-drive into /opt/icloud-drive/config, added /opt/icloud-drive/local and added the volume to the compose file. All works great. And the keyring does indeed persist. Awesome!

Once this PR is merged into the main repo, all solved as far as I'm concerned! 

Thanks!
